x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheServiceTests.java

```diff
@@ -221,44 +226,150 @@ public void testPut() throws Exception {
-        cacheFile.release(evictionListener);
-
-        cacheService.runIfShardMarkedAsEvictedInCache(
-            snapshotId,
-            indexId,
-            shardId,
-            () -> { assert false : "should not be called: shard eviction marker is removed"; }
+    }
+
+    public void testProcessShardEviction() throws Exception {
+        final CacheService cacheService = defaultCacheService();
+        cacheService.start();
+
+        final List<CacheFile> randomCacheFiles = randomCacheFiles(cacheService);
+        assertThat(cacheService.pendingShardsEvictions(), aMapWithSize(0));
+
+        final ShardEviction shard = randomShardEvictionFrom(randomCacheFiles);
+        final List<CacheFile> cacheFilesAssociatedWithShard = filterByShard(shard, randomCacheFiles);
+        cacheFilesAssociatedWithShard.forEach(cacheFile -> assertTrue(Files.exists(cacheFile.getFile())));
+
+        final BlockingEvictionListener blockingListener = new BlockingEvictionListener();
+        final CacheFile randomCacheFile = randomFrom(cacheFilesAssociatedWithShard);
+        assertTrue(Files.exists(randomCacheFile.getFile()));
+        randomCacheFile.acquire(blockingListener);
+
+        cacheService.markShardAsEvictedInCache(shard.getSnapshotUUID(), shard.getSnapshotIndexName(), shard.getShardId());
+
+        final Map<CacheFile, Boolean> afterShardRecoveryCacheFiles = ConcurrentCollections.newConcurrentMap();
+        final Thread shardRecoveryThread = new Thread(() -> {
+            cacheService.waitForCacheFilesEvictionIfNeeded(shard.getSnapshotUUID(), shard.getSnapshotIndexName(), shard.getShardId());
+            for (CacheFile cacheFile : cacheFilesAssociatedWithShard) {
+                afterShardRecoveryCacheFiles.put(cacheFile, Files.exists(cacheFile.getFile()));
+            }
+        });
+        shardRecoveryThread.start();
+
+        blockingListener.waitForBlock();
+
+        final Map<ShardEviction, Future<?>> pendingShardsEvictions = cacheService.pendingShardsEvictions();
+        assertTrue(cacheService.isPendingShardEviction(shard));
+        assertThat(pendingShardsEvictions, aMapWithSize(1));
+
+        final Future<?> shardEvictionFuture = pendingShardsEvictions.get(shard);
+        assertTrue(Files.exists(randomCacheFile.getFile()));
+        assertThat(shardEvictionFuture, notNullValue());
+        assertFalse(shardEvictionFuture.isDone());
+
+        blockingListener.unblock();
+
+        FutureUtils.get(shardEvictionFuture);
+        shardRecoveryThread.join();
+
+        cacheFilesAssociatedWithShard.forEach(
+            cacheFile -> assertFalse("Cache file should be evicted: " + cacheFile, Files.exists(cacheFile.getFile()))
         );
-        cacheService.close();
+        afterShardRecoveryCacheFiles.forEach(
+            (cacheFile, exists) -> assertFalse(
+                "Cache file should have been evicted after shard recovery: " + cacheFile,
+                Files.exists(cacheFile.getFile())
```

Comment:
Should this be:

```suggestion
                exists
```

otherwise `exists` looks unused?

Reply:
Yes, that's a wrong copy/paste. I pushed 5293eca6db7cc7d4c88d01c4c306a76f48ad968c